File: packages/hap-python/hap/verify.py
"""
HAP claim verification functions
"""

import json
from datetime import datetime, timezone
from typing import Any, Optional
from urllib.parse import urlparse

import httpx
import jwt
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey
from cryptography.hazmat.primitives.serialization import Encoding, PublicFormat

from hap.types import (
    HAP_ID_REGEX,
    HapClaim,
    HapJwk,
    HapWellKnown,
    VerificationResponse,
)


def is_valid_hap_id(hap_id: str) -> bool:
    """
    Validates a HAP ID format.

    Args:
        hap_id: The HAP ID to validate

    Returns:
        True if the ID matches the format hap_[a-zA-Z0-9]{12}
    """
    return bool(HAP_ID_REGEX.match(hap_id))


async def fetch_public_keys(
    issuer_domain: str,
    timeout: float = 10.0,
    client: Optional[httpx.AsyncClient] = None,
) -> HapWellKnown:
    """
    Fetches the public keys from a VA's well-known endpoint.

    Args:
        issuer_domain: The VA's domain (e.g., "ballista.app")
        timeout: Request timeout in seconds
        client: Optional httpx client to use

    Returns:
        The VA's public key configuration
    """
    url = f"https://{issuer_domain}/.well-known/hap.json"

    if client:
        response = await client.get(url, timeout=timeout)
    else:
        async with httpx.AsyncClient() as c:
            response = await c.get(url, timeout=timeout)

    response.raise_for_status()
    return response.json()


async def fetch_claim(
    hap_id: str,
    issuer_domain: str,
    timeout: float = 10.0,
    client: Optional[httpx.AsyncClient] = None,
) -> VerificationResponse:
    """
    Fetches and verifies a HAP claim from a VA.

    Args:
        hap_id: The HAP ID to verify
        issuer_domain: The VA's domain (e.g., "ballista.app")
        timeout: Request timeout in seconds
        client: Optional httpx client to use

    Returns:
        The verification response from the VA
    """
    if not is_valid_hap_id(hap_id):
        return {"valid": False, "error": "invalid_format"}

    url = f"https://{issuer_domain}/api/v1/verify/{hap_id}"

    if client:
        response = await client.get(url, timeout=timeout)
    else:
        async with httpx.AsyncClient() as c:
            response = await c.get(url, timeout=timeout)

    return response.json()


def _base64url_decode(data: str) -> bytes:
    """Decode base64url string to bytes."""
    import base64

    # Add padding if needed
    padding = 4 - len(data) % 4
    if padding != 4:
        data += "=" * padding
    return base64.urlsafe_b64decode(data)


def _jwk_to_public_key(jwk: HapJwk) -> Ed25519PublicKey:
    """Convert JWK to Ed25519 public key."""
    from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey

    x_bytes = _base64url_decode(jwk["x"])
    return Ed25519PublicKey.from_public_bytes(x_bytes)


async def verify_signature(
    jws: str,
    issuer_domain: str,
    timeout: float = 10.0,
    client: Optional[httpx.AsyncClient] = None,
) -> dict[str, Any]:
    """
    Verifies a JWS signature against a VA's public keys.

    Args:
        jws: The JWS compact serialization string
        issuer_domain: The VA's domain to fetch public keys from
        timeout: Request timeout in seconds
        client: Optional httpx client to use

    Returns:
        Dict with 'valid' boolean, 'claim' if valid, 'error' if invalid
    """
    try:
        # Fetch public keys from the VA
        well_known = await fetch_public_keys(issuer_domain, timeout, client)

        # Parse the JWS header to get the key ID
        header = jwt.get_unverified_header(jws)
        kid = header.get("kid")

        if not kid:
            return {"valid": False, "error": "JWS header missing kid"}

        # Find the matching key
        jwk = next((k for k in well_known["keys"] if k["kid"] == kid), None)
        if not jwk:
            return {"valid": False, "error": f"Key not found: {kid}"}

        # Convert JWK to public key
        public_key = _jwk_to_public_key(jwk)

        # Verify the signature using PyJWT
        claim = jwt.decode(
            jws,
            public_key,
            algorithms=["EdDSA"],
            options={"verify_aud": False},
        )

        # Verify the issuer matches
        if claim.get("iss") != issuer_domain:
            return {
                "valid": False,
                "error": f"Issuer mismatch: expected {issuer_domain}, got {claim.get('iss')}",
            }

        return {"valid": True, "claim": claim}

    except Exception as e:
        return {"valid": False, "error": str(e)}


async def verify_hap_claim(
    hap_id: str,
    issuer_domain: str,
    verify_sig: bool = True,
    timeout: float = 10.0,
    client: Optional[httpx.AsyncClient] = None,
) -> Optional[HapClaim]:
    """
    Fully verifies a HAP claim: fetches from VA and optionally verifies signature.

    Args:
        hap_id: The HAP ID to verify
        issuer_domain: The VA's domain
        verify_sig: Whether to verify the cryptographic signature
        timeout: Request timeout in seconds
        client: Optional httpx client to use

    Returns:
        The claim if valid, None if not found or invalid
    """
    # Fetch the claim from the VA
    response = await fetch_claim(hap_id, issuer_domain, timeout, client)

    # Check if valid
    if not response.get("valid"):
        return None

    # Optionally verify the signature
    if verify_sig and "jws" in response:
        sig_result = await verify_signature(response["jws"], issuer_domain, timeout, client)
        if not sig_result.get("valid"):
            return None

    return response.get("claims")


def extract_hap_id_from_url(url: str) -> Optional[str]:
    """
    Extracts the HAP ID from a verification URL.

    Args:
        url: The verification URL (e.g., "https://ballista.app/v/hap_abc123xyz456")

    Returns:
        The HAP ID or None if not found
    """
    try:
        parsed = urlparse(url)
        path_parts = parsed.path.split("/")
        last_part = path_parts[-1] if path_parts else ""

        if is_valid_hap_id(last_part):
            return last_part

        return None
    except Exception:
        return None


def is_claim_expired(claim: HapClaim) -> bool:
    """
    Checks if a claim is expired.

    Args:
        claim: The HAP claim to check

    Returns:
        True if the claim has an exp field and is expired
    """
    exp = claim.get("exp")
    if not exp:
        return False

    exp_date = datetime.fromisoformat(exp.replace("Z", "+00:00"))
    return exp_date < datetime.now(timezone.utc)


def is_claim_for_company(claim: HapClaim, company_domain: str) -> bool:
    """
    Checks if the claim target matches the expected company.

    Args:
        claim: The HAP claim to check
        company_domain: The expected company domain

    Returns:
        True if the claim's target domain matches
    """
    claim_type = claim.get("type")

    if claim_type == "human_effort":
        to = claim.get("to", {})
        return to.get("domain") == company_domain

    if claim_type == "employer_commitment":
        employer = claim.get("employer", {})
        return employer.get("domain") == company_domain

    return False

File: packages/hap-js/src/verify.ts
/**
 * HAP claim verification functions
 */

import * as jose from "jose";
import {
  HAP_ID_REGEX,
  HapClaim,
  HapJwk,
  HapWellKnown,
  SignatureVerificationResult,
  VerificationResponse,
  VerifyOptions,
} from "./types";

/**
 * Validates a HAP ID format
 * @param id - The HAP ID to validate
 * @returns true if the ID matches the format hap_[a-zA-Z0-9]{12}
 */
export function isValidHapId(id: string): boolean {
  return HAP_ID_REGEX.test(id);
}

/**
 * Fetches the public keys from a VA's well-known endpoint
 * @param issuerDomain - The VA's domain (e.g., "ballista.io")
 * @param options - Optional configuration
 * @returns The VA's public key configuration
 */
export async function fetchPublicKeys(
  issuerDomain: string,
  options: VerifyOptions = {},
): Promise<HapWellKnown> {
  const fetchFn = options.fetch ?? fetch;
  const timeout = options.timeout ?? 10000;

  const controller = new AbortController();
  const timeoutId = setTimeout(() => controller.abort(), timeout);

  try {
    const url = `https://${issuerDomain}/.well-known/hap.json`;
    const response = await fetchFn(url, {
      signal: controller.signal,
      headers: {
        Accept: "application/json",
      },
    });

    if (!response.ok) {
      throw new Error(`Failed to fetch public keys: HTTP ${response.status}`);
    }

    return (await response.json()) as HapWellKnown;
  } finally {
    clearTimeout(timeoutId);
  }
}

/**
 * Fetches and verifies a HAP claim from a VA
 * @param hapId - The HAP ID to verify
 * @param issuerDomain - The VA's domain (e.g., "ballista.io")
 * @param options - Optional configuration
 * @returns The verification response from the VA
 */
export async function fetchClaim(
  hapId: string,
  issuerDomain: string,
  options: VerifyOptions = {},
): Promise<VerificationResponse> {
  if (!isValidHapId(hapId)) {
    return { valid: false, error: "invalid_format" };
  }

  const fetchFn = options.fetch ?? fetch;
  const timeout = options.timeout ?? 10000;

  const controller = new AbortController();
  const timeoutId = setTimeout(() => controller.abort(), timeout);

  try {
    const url = `https://${issuerDomain}/api/v1/verify/${hapId}`;
    const response = await fetchFn(url, {
      signal: controller.signal,
      headers: {
        Accept: "application/json",
      },
    });

    return (await response.json()) as VerificationResponse;
  } finally {
    clearTimeout(timeoutId);
  }
}

/**
 * Verifies a JWS signature against a VA's public keys
 * @param jws - The JWS compact serialization string
 * @param issuerDomain - The VA's domain to fetch public keys from
 * @param options - Optional configuration
 * @returns Verification result with decoded claim if valid
 */
export async function verifySignature(
  jws: string,
  issuerDomain: string,
  options: VerifyOptions = {},
): Promise<SignatureVerificationResult> {
  try {
    // Fetch public keys from the VA
    const wellKnown = await fetchPublicKeys(issuerDomain, options);

    // Parse the JWS header to get the key ID
    const header = jose.decodeProtectedHeader(jws);
    const kid = header.kid;

    if (!kid) {
      return { valid: false, error: "JWS header missing kid" };
    }

    // Find the matching key
    const jwk = wellKnown.keys.find((k: HapJwk) => k.kid === kid);
    if (!jwk) {
      return { valid: false, error: `Key not found: ${kid}` };
    }

    // Import the public key
    const publicKey = await jose.importJWK(jwk, "EdDSA");

    // Verify the signature
    const { payload } = await jose.compactVerify(jws, publicKey);

    // Decode the payload
    const claim = JSON.parse(new TextDecoder().decode(payload)) as HapClaim;

    // Verify the issuer matches
    if (claim.iss !== issuerDomain) {
      return {
        valid: false,
        error: `Issuer mismatch: expected ${issuerDomain}, got ${claim.iss}`,
      };
    }

    return { valid: true, claim };
  } catch (error) {
    const message = error instanceof Error ? error.message : "Unknown error";
    return { valid: false, error: message };
  }
}

/**
 * Fully verifies a HAP claim: fetches from VA and optionally verifies signature
 * @param hapId - The HAP ID to verify
 * @param issuerDomain - The VA's domain
 * @param options - Optional configuration
 * @returns The claim if valid, null if not found or invalid
 */
export async function verifyHapClaim(
  hapId: string,
  issuerDomain: string,
  options: VerifyOptions & { verifySignature?: boolean } = {},
): Promise<HapClaim | null> {
  // Fetch the claim from the VA
  const response = await fetchClaim(hapId, issuerDomain, options);

  // Check if valid
  if (!response.valid) {
    return null;
  }

  // Optionally verify the signature (recommended for high-security use cases)
  if (options.verifySignature !== false && "jws" in response) {
    const sigResult = await verifySignature(
      response.jws,
      issuerDomain,
      options,
    );
    if (!sigResult.valid) {
      return null;
    }
  }

  return "claims" in response ? response.claims : null;
}

/**
 * Extracts the HAP ID from a verification URL
 * @param url - The verification URL (e.g., "https://ballista.io/v/hap_abc123xyz456")
 * @returns The HAP ID or null if not found
 */
export function extractHapIdFromUrl(url: string): string | null {
  try {
    const parsed = new URL(url);
    const pathParts = parsed.pathname.split("/");
    const lastPart = pathParts[pathParts.length - 1];

    if (isValidHapId(lastPart)) {
      return lastPart;
    }

    return null;
  } catch {
    return null;
  }
}

/**
 * Checks if a claim is expired
 * @param claim - The HAP claim to check
 * @returns true if the claim has an exp field and is expired
 */
export function isClaimExpired(claim: HapClaim): boolean {
  if (!claim.exp) {
    return false;
  }

  const expDate = new Date(claim.exp);
  return expDate < new Date();
}

/**
 * Checks if the claim target matches the expected company
 * @param claim - The HAP claim to check
 * @param companyDomain - The expected company domain
 * @returns true if the claim's target domain matches
 */
export function isClaimForCompany(
  claim: HapClaim,
  companyDomain: string,
): boolean {
  if (claim.type === "human_effort") {
    return claim.to.domain === companyDomain;
  }
  if (claim.type === "employer_commitment") {
    return claim.employer.domain === companyDomain;
  }
  return false;
}
